### ISSUE TITLE
log rollover fails on Windows.

### ISSUE THREAD

**Issue by petersilva** [MetPX/sarracenia#785]
user reported complaint using the self-extracting windows executable, 3.0.44
```
Traceback (most recent call last):
  File "logging\handlers.py", line 74, in emit
  File "C:\Program Files\metpx-sr3\pkgs\sarracenia\instance.py", line 32, in doRollover
    super().doRollover()
  File "logging\handlers.py", line 435, in doRollover
  File "logging\handlers.py", line 115, in rotate
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\Administrator\\AppData\\Local\\MetPX\\sr3\\Cache\\log\\subscribe_cbc_01.log' -> 'C:\\Users\\Administrator\\AppData\\Local\\MetPX\\sr3\\Cache\\log\\subscribe_cbc_01.log.2023-10-03'
```

. The log file rollover is not working. So the file keeps growing until I manually stop the service and rename it.

**Comment by petersilva** [comment]
I have implemented some work for #784 should help here.
with this patch: https://github.com/MetPX/sarracenia/commit/a6851e719c84e22c7b7454701843bb4e610d166f

I can set up log rollover to be every 30 seconds, and can test.

**Comment by petersilva** [comment]
With the v03_issue784 branch, I set the log to rotate every 30 seconds.  It works perfectly on linux.  On Windows:
* it doesn't rotate, but doesn't crash immediately... after like a few minutes, it fails with the error message reported about threading.

So... it's easy to reproduce... but... why does it think two processes are accessing the file? they aren't... fixing it is not immediately obvious.  I tried asking ChatGPT... it said: don't use multiple processes... ok, but I'm not.  other suggestion: use another python module called loguru, which apparently does a better job.  Bot loguru is semantically different, and won't do what the current logging does.

so...  no obvious solution so far...

**Comment by petersilva** [comment]
Here is this is what we do on linux:
* we use the logging.handlers.TimedRotatingFileHandler to get log messages re-directed to a file.
* we use os.dup2() to re-direct stdout and stderr onto the log file.  This is done using dup2 because we don't only want log messages re-directed, or even print() messages redirected, we want the stdout/stderr of sub-processes we launch to be re-directed into the log as well.  Just re-directing python's idea of stdout does not do that.
* we subclassed the handler so that we move the re-direction to the new file after each rotation.
* in sr3, we use and value the ability to tune the logging level within each module. One can turn on "debug" for a particular plugin, or moth.amqp.AMQP and set the logging message threshold to "info" or "warning" in the rest of the application.

This works on linux. by which is meant:
* stdout and stderr are re-directed to the log file, including sub-processes.
* log rollover every 10 seconds seems to work, and there is no trailing output in the old log file (re-direction switches to new file.)
* varying log levels are respected.
* all writing to the log is being done by one process, and only one process has the file open, unless a sub-process has been spawned (in a plugin) and we are waiting for it.

method:
when calling start, the entry point launches a sub-process.  The parent process does this:
```
                with open(lfn, "a") as lf:
                    subprocess.Popen(cmd,
                                 stdin=subprocess.DEVNULL,
                                 stdout=lf,
                                 stderr=subprocess.STDOUT)

```
so the parent calls subprocess.Popen with the opened lfn (aka: log file name) in append mode, in a with clause, ensuring that as soon as the with clause exits, the file will be closed, and the with clause exits once the Popen returns, which is immediately after launching the sub-process.

So the parent does not have the log file open.  There is no second process with the file open.


On Windows: 
* when dup2 is called, it breaks standard output, and the next time standard output is used, results in a crash... traceback something like this:

```
FIXME: TBC

```


* as a result of the above dup2 is not called on windows, so there it is unclear how to get the re-direction of python to switch to the new file after log rotation occurs.

**Comment by petersilva** [comment]
example code:
```python
import logging
import logging.handlers
import os
import sys
import time


logger = logging.getLogger(__name__)

log_format = '%(asctime)s [%(levelname)s] %(name)s %(funcName)s %(message)s'

logger.info("hithere" )
logfilename = "lrh.log"

handler = logging.handlers.TimedRotatingFileHandler(
    logfilename,
    when='S',
    interval=15,
    backupCount=4)
handler.setFormatter(logging.Formatter(log_format))

logger.addHandler(handler)
logger.setLevel(logging.DEBUG)


print("redirect std files to log")
if True:
    os.dup2(handler.stream.fileno(), 1)
    # normally dup2 both, but won't see the error if I do that here.
    #os.dup2(handler.stream.fileno(), 2)

while True:
    logger.info("writing")
    time.sleep(4)
    print( f"{time.time()} awake!")
```


reproducing the crash with really small example:
```

PS C:\Users\SilvaP2\AppData\Local\MetPX\sr3\Cache\log> python3 lrh.py
redirect std files to log
--- Logging error ---
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\logging\handlers.py", line 74, in emit
    self.doRollover()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\logging\handlers.py", line 435, in doRollover
    self.rotate(self.baseFilename, dfn)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\logging\handlers.py", line 115, in rotate
    os.rename(source, dest)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\SilvaP2\\AppData\\Local\\MetPX\\sr3\\Cache\\log\\lrh.log' -> 'C:\\Users\\SilvaP2\\AppData\\Local\\MetPX\\sr3\\Cache\\log\\lrh.log.2023-10-22_15-03-41'
Call stack:
  File "C:\Users\SilvaP2\AppData\Local\MetPX\sr3\Cache\log\lrh.py", line 33, in <module>
    logger.info("writing")
Message: 'writing'
Arguments: ()
Traceback (most recent call last):
  File "C:\Users\SilvaP2\AppData\Local\MetPX\sr3\Cache\log\lrh.py", line 35, in <module>
    print( f"{time.time()} awake!")
OSError: [WinError 6] The handle is invalid
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
OSError: [WinError 6] The handle is invalid
PS C:\Users\SilvaP2\AppData\Local\MetPX\sr3\Cache\log>

'''

that code crashes in different ways at different times... after a few invocations, it produces the "being used by another process" error... but after a thorough search... nope, there is no other process.  So the message is wrong/misleading.

**Comment by petersilva** [comment]
light reading:
* https://stackoverflow.com/questions/29167596/python-os-dup2-redirect-enables-output-buffering-on-windows-python-consoles (nothing there works... get *I/o on a closed file* )
* I think what the "another process" thing is about is just have the file opened twice (either two python file structures, or two file descriptors for the same file) within the same process.
* This is kind of hard to avoid if redirecting stdout, stderr and log to a single file... that's three descriptors right there.

**Comment by petersilva** [comment]
tried a custom rollover that closes the file before, and uses dup to reopen after:

```python

class RedirectedTimedRotatingFileHandler(
        logging.handlers.TimedRotatingFileHandler):
    def doRollover(self):

        #os.close(1)
        #os.close(2)
        old_stdout = sys.stdout
        old_stdout_fd = sys.stdout.fileno()
        os.close(1)
        #logger.critical( f" prior to rotation: stdout: {old_stdout}  fd: {old_stdout_fd}")

        super().doRollover()

        #sys.stderr=self.stream
        os.dup2(self.stream.fileno(), old_stdout_fd)
        #os.dup2(self.stream.fileno(), 2)


```

no joy:

```

PS C:\Users\SilvaP2\AppData\Local\MetPX\sr3\Cache\log> python3 lrh.py
redirect std files to log
Traceback (most recent call last):
  File "C:\Users\SilvaP2\AppData\Local\MetPX\sr3\Cache\log\lrh.py", line 55, in <module>
    print( f"{time.time()} awake!")
ValueError: I/O operation on closed file.
PS C:\Users\SilvaP2\AppData\Local\MetPX\sr3\Cache\log>
````

**Comment by petersilva** [comment]
Ok... So I don't have amy solution in view... to have sr3 rotate logs on windows the same way it does on linux.  It is time to think about work-arounds. 

Set the log rotation to disable rotation for now... and use other windows methods... so
when you want to rotate the logs:  
```

logRotateInterval 7d

```

need to set up a periodically run task on Windows to run this:

```

sr3 stop 
sr3_win_rotateLogs.py 4 
sr3 start

```

The 4 refers to how many logs to keep.  The scheduling of the tasks will determine the frequency of rollover.

One can grab the sr3_win_rotateLogs.py from the commit... https://github.com/MetPX/sarracenia/commit/c3ec22c55d4f09ee8765d9dd7d7772f9b5e4ff56

**Comment by petersilva** [comment]
todo:
* add the sr3_win_rotateLogs.py entry point to setup.py, conditional on being on windows.
* Add some documentation? (perhaps a caveat in the documentation of *logRotate* optons
* perhaps document how to set up a routine windows Task ( I think there is already documentation about that for sr3 sanity (or v2: sr audit.)

Once that is done... can say we have a work-around.

**Comment by petersilva** [comment]
branch for work-around is: https://github.com/MetPX/sarracenia/compare/v03_issue785_workaround

**Comment by petersilva** [comment]
#716 as well as this would be resolved by explicitly capturing stdout/stderr of subprocesses. ... a bit of a project though.
